### PR TITLE
Login functionality

### PR DESCRIPTION
**Pull request by camillebaldock**

### CODE REVIEW

File: service/server.py [hunk 1]
```diff
@@ -1,39 +1,119 @@
...
+import requests
+from wtforms.fields import StringField, PasswordField
+from wtforms.validators import Required, Length
+
+from service import app, login_manager
+
+
+REGISTER_TITLE_API = app.config['REGISTER_TITLE_API']
+LOGIN_JSON = '{{"credentials":{{"user_id":"{}","password":"{}"}}}}'
+FORWARD_SLASH = '/'
```
Comment (camillebaldock) [inline service/server.py]: don't think this is being used anywhere?

File: service/server.py [hunk 2]
```diff
@@ -1,39 +1,119 @@
...
+@app.route('/login', methods=['POST'])
+def signin():
+    form = SigninForm(csrf_enabled=_is_csrf_enabled())
+    if not form.validate():
+        # entered details from login form incorrectly so send back to same page with form error messages
+        return render_template('display_login.html', asset_path='../static/', form=form)
+    else:
+        username = form.username.data
+        # form has correct details. Now need to check authorisation
+        authorised = LOGIN_API_CLIENT.authenticate_user(username, form.password.data)
```
Comment (camillebaldock) [inline service/server.py:98]: we seem to be passing through parameters to the API exactly as the user enters them here, what would happen if I enter a json-like structure in the fields that would make the JSON being sent to the login-api invalid?

File: service/server.py [hunk 3]
```diff
@@ -147,13 +234,32 @@ def get_address_lines(address_data):
...
+class SigninForm(Form):
+    username = StringField('username', [Required(message='Username is required'), Length(min=4, max=70, message='Username is incorrect')])
+    password = PasswordField('password', [Required(message='Password is required')])
+    
+    def __init__(self, *args, **kwargs):
+        Form.__init__(self, *args, **kwargs)
+
+
+class TitleSearchForm(Form):
+    pass
```
Comment (camillebaldock) [inline service/server.py:278]: Could you please comment a bit on this? I am not confident I understand what this does. Does this mean we are happy with whatever field content is entered for title search? and this was added just to have the hidden tag helper?

File: service/server.py (hunk 1 above)
Comment (lgonczar) [inline service/server.py]: Don't think it should even be a constant. Will remove.

File: service/server.py (hunk 3 above)
Comment (lgonczar) [inline service/server.py:278]: This was added for the hidden_tag helper. I also left a TODO saying we should use the WTF form for HTTP form processing and rendering. I'll add a line of comment.

File: service/server.py [hunk 4]
```diff
@@ -1,39 +1,138 @@
...
+        return render_template('display_login.html', asset_path='../static/', form=form)
+
+    next_url = request.args.get('next', 'title-search')
+
+    # form was valid
+    username = form.username.data
+    too_many_bad_logins = BAD_LOGIN_COUNTER[username] > MAX_LOGIN_ATTEMPTS
+    if not too_many_bad_logins:
+        # form has correct details. Now need to check authorisation
+        authorised = LOGIN_API_CLIENT.authenticate_user(username, form.password.data)
```
Comment (camillebaldock) [inline service/server.py:98]: we seem to be passing through parameters to the API exactly as the user enters them here, what would happen if I enter a json-like structure in the fields that would make the JSON being sent to the login-api invalid?

Comment (lgonczar) [inline service/server.py:98]: Good spot, fixed

**Pull request merged**